tests/components/eafm/test_sensor.py

```diff
@@ -0,0 +1,431 @@
+    assert state.state == "5"
+
+
+@pytest.mark.parametrize("exception", CONNECTION_EXCEPTIONS)
+async def test_reading_unavailable(hass, mock_get_station, exception):
+    """Test that a sensor is marked as unavailable if there is a connection error."""
+    _, poll = await async_setup_test_fixture(
+        hass,
+        mock_get_station,
+        {
+            "label": "My station",
+            "measures": [
+                {
+                    "@id": "really-long-unique-id",
+                    "label": "York Viking Recorder - level-stage-i-15_min----",
+                    "qualifier": "Stage",
+                    "parameterName": "Water Level",
+                    "latestReading": {"value": 5},
+                    "stationReference": "L1234",
+                    "unit": "http://qudt.org/1.1/vocab/unit#Meter",
+                    "unitName": "m",
+                }
+            ],
+        },
+    )
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "5"
+
+    await poll(exception)
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "unavailable"
+
+
+@pytest.mark.parametrize("exception", CONNECTION_EXCEPTIONS)
+async def test_recover_from_failure(hass, mock_get_station, exception):
+    """Test that a sensor recovers from failures."""
+    _, poll = await async_setup_test_fixture(
+        hass,
+        mock_get_station,
+        {
+            "label": "My station",
+            "measures": [
+                {
+                    "@id": "really-long-unique-id",
+                    "label": "York Viking Recorder - level-stage-i-15_min----",
+                    "qualifier": "Stage",
+                    "parameterName": "Water Level",
+                    "latestReading": {"value": 5},
+                    "stationReference": "L1234",
+                    "unit": "http://qudt.org/1.1/vocab/unit#Meter",
+                    "unitName": "m",
+                }
+            ],
+        },
+    )
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "5"
+
+    await poll(exception)
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "unavailable"
+
+    await poll(
+        {
+            "label": "My station",
+            "measures": [
+                {
+                    "@id": "really-long-unique-id",
+                    "label": "York Viking Recorder - level-stage-i-15_min----",
+                    "qualifier": "Stage",
+                    "parameterName": "Water Level",
+                    "latestReading": {"value": 56},
+                    "stationReference": "L1234",
+                    "unit": "http://qudt.org/1.1/vocab/unit#Meter",
+                    "unitName": "m",
+                }
+            ],
+        },
+    )
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "56"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)